src/disco/store/fd_pending_slots.c

```diff
@@ -13,27 +14,25 @@ fd_pending_slots_new( void * mem, ulong lo_wmark ) {
     return NULL;
   }
 
-  ulong footprint = fd_pending_slots_footprint();
+  fd_memset( mem, 0, fd_pending_slots_footprint() );
 
-  fd_memset( mem, 0, footprint );
-  ulong laddr = (ulong)mem;
-  fd_pending_slots_t * pending_slots = (void *)laddr;
-  pending_slots->lo_wmark = lo_wmark;
-  pending_slots->start = 0;
-  pending_slots->end = 0;
-  pending_slots->lock = 0;
+  fd_pending_slots_t * pending_slots = mem;
+  ulong                laddr         = (ulong)(mem) + sizeof( fd_pending_slots_t );
 
-  laddr += sizeof( fd_pending_slots_t );
-  pending_slots->pending = (void *)laddr;
+  laddr                = fd_ulong_align_up( laddr, fd_pending_slots_treap_align() );
+  pending_slots->treap = fd_pending_slots_treap_join( fd_pending_slots_treap_new( (void *)(laddr), FD_BLOCK_MAX ) );
+  laddr               += fd_pending_slots_treap_footprint( FD_BLOCK_MAX );
 
-  laddr += sizeof(long) * FD_PENDING_MAX;
+  laddr                = fd_ulong_align_up( laddr, fd_pending_slots_pool_align() );
+  pending_slots->pool  = fd_pending_slots_pool_join( fd_pending_slots_pool_new( (void*)(laddr), FD_BLOCK_MAX ) );
+  laddr               += fd_pending_slots_pool_footprint( FD_BLOCK_MAX );
 
-  FD_TEST( laddr == (ulong)mem + footprint );
+  FD_TEST( laddr == (ulong)mem + fd_pending_slots_footprint() );
 
+  fd_pending_slots_treap_seed( pending_slots->pool, FD_BLOCK_MAX, seed );
```

Comment:
It is unclear if we should generate a random number on re-insert. For correctness it is fine but for (computational) complexity maybe not.